LIVE-2653: Follow icon

faresite opened the pull request:
## Why are you doing this?

This icon should go into source however as a quick solution the svg is in our codebase.

https://theguardian.atlassian.net/browse/LIVE-2653

@benwuersching @prisalcalde 


## Changes

- Add Follow icon svg
- icon styling

## Screenshots

![Screenshot 2021-06-24 at 11 57 05](https://user-images.githubusercontent.com/12860328/123252717-82fa3600-d4e4-11eb-8ab5-7f98efb77ff8.png)
![Screenshot 2021-06-24 at 11 57 00](https://user-images.githubusercontent.com/12860328/123252720-8392cc80-d4e4-11eb-92ff-eb0a395b0add.png)
![Screenshot 2021-06-24 at 11 56 46](https://user-images.githubusercontent.com/12860328/123252723-842b6300-d4e4-11eb-8d8a-68bf8cd793d5.png)
![Screenshot 2021-06-24 at 11 56 41](https://user-images.githubusercontent.com/12860328/123252725-842b6300-d4e4-11eb-98df-7a69e2fdd081.png)
![Screenshot 2021-06-24 at 11 56 29](https://user-images.githubusercontent.com/12860328/123252728-84c3f980-d4e4-11eb-8547-82c65d4c47eb.png)
![Screenshot 2021-06-24 at 11 56 22](https://user-images.githubusercontent.com/12860328/123252729-84c3f980-d4e4-11eb-8f79-870e9043d500.png)
![Screenshot 2021-06-24 at 11 56 07](https://user-images.githubusercontent.com/12860328/123252731-84c3f980-d4e4-11eb-90b5-9499af6489ac.png)
![Screenshot 2021-06-24 at 11 55 57](https://user-images.githubusercontent.com/12860328/123252732-855c9000-d4e4-11eb-8d7f-6c4481f05840.png)

Commit:
add follow iconb

JamieB-gu commented on src/components/follow.tsx:
```diff
@@ -25,6 +25,28 @@ const styles = (kicker: string, inverted: string): SerializedStyles => css`
 	margin-left: 0;
+	margin-top: 4px;
+
+	svg {
```
Perhaps worth considering a `css` attribute on the `<svg>` over type selectors? Not a blocker.

JamieB-gu commented on src/components/follow.tsx:
```diff
@@ -25,6 +25,28 @@ const styles = (kicker: string, inverted: string): SerializedStyles => css`
+	margin-top: 4px;
+
+	svg {
+		width: 24px;
+		height: 24px;
+		margin-bottom: -6px;
```
Worth converting these to `rem` for scaling reasons (the scale slider)?

faresite replied:
I checked with the templates (that we're trying to mimic) and they don't scale with the slider so decided to go with px. What do you think?

JamieB-gu replied:
We're not trying to mimic the scaling behaviour of the templates. They _only_ scale body text. We scale everything (with the exception of the headline).

JamieB-gu commented on src/components/follow.tsx:
```diff
@@ -25,6 +25,28 @@ const styles = (kicker: string, inverted: string): SerializedStyles => css`
+			fill: ${inverted};
+		}
+		path {
+			fill: #121212;
```
This is `neutral.7` in Source. I'm not sure there's a semantic colour name (`background.inverse` doesn't really work for SVG paths 😬).

JamieB-gu commented on src/components/follow.tsx:
```diff
@@ -25,6 +25,28 @@ const styles = (kicker: string, inverted: string): SerializedStyles => css`
 	border: none;
 	background: none;
 	margin-left: 0;
+	margin-top: 4px;
```
`remSpace[1]`?

JamieB-gu reviewed (commented):
👍 Few small comments.

Commit:
address pr review

JamieB-gu commented on src/components/follow.tsx:
```diff
@@ -59,16 +44,42 @@ function getStyles({ theme }: Format): SerializedStyles {
+	const iconStyles = ({ theme }: Format): SerializedStyles => {
+		const { kicker, inverted } = getThemeStyles(theme);
+		return css`
+			width: ${remSpace[6]};
+			height: ${remSpace[6]};
```
Technically these aren't spaces but it probably doesn't matter too much 🙂

JamieB-gu reviewed (commented):
There's one `margin-top` that isn't yet resolved, but apart from that looks good.

benwuersching commented:
@faresite @JamieB-gu Great to see we're fixing this! I just wanted to make sure we're aware that this component comes with two states/icons, i.e. the "not following" state and the "following" state. The two icons are subtly different.
![follow_unfollow](https://user-images.githubusercontent.com/23035818/123287192-7a1a5c00-d506-11eb-8107-a5b5f7a6a487.png)

Commit:
following state icon

faresite commented:
Thanks @benwuersching for pointing this out. I've now added the following state, here's a gif in light and dark mode.

![2021-06-24 18 19 58](https://user-images.githubusercontent.com/12860328/123306320-0f265080-d519-11eb-883d-4852bbf2a295.gif)
![2021-06-24 18 19 29](https://user-images.githubusercontent.com/12860328/123306344-13eb0480-d519-11eb-9ca3-9af3a6eb5cc4.gif)

Commit:
remify

Commit:
update snaps

prisalcalde commented:
Great @faresite ! Looking good to me.

benwuersching commented:
Thank you @faresite !

Commit:
followIcon component

Commit:
split out component to be hydrated

Commit:
render FollowStatus in client

Commit:
update snaps

faresite commented on webpack.config.ts:
```diff
@@ -64,6 +64,7 @@ function resolve(
 	// We rely on Buffer for our bridget thrift client
 	if (isClient) {
 		resolveOptions.alias = { ...resolveOptions.alias, Buffer: 'buffer' };
+		resolveOptions.fallback = { crypto: false };
```
@JamieB-gu webpack stopped compiling client side bundles out of the blue, don't know why .. it was something to do with `image-rendering` and `crypto`.. but I don't understand it as nothing has changed 🤷‍♂️

JamieB-gu replied:
Something has changed, we've added a new component to the `import`s in the client-side bundle. And the import tree results in `image-rendering` being included client-side. I'd suggest breaking `FollowStatus` out into its own file rather than exporting it from `Follow`. That should narrow the import tree enough to exclude `image-rendering`.

JamieB-gu committed:
Extract `FollowStatus` into a separate file

- Remove webpack `resolveOptions` config for crypto
- Imported `FollowStatus` from new file in client bundle
- Extracted `FollowStatus` from `follow.tsx`
- Simplified `follow.tsx`
- Added `followStatus.tsx`

JamieB-gu committed:
Lint fix

JamieB-gu commented on src/components/follow.tsx:
```diff
@@ -37,6 +44,79 @@ function getStyles({ theme }: Format): SerializedStyles {
 
+const FollowIcon: FC<IconProps> = ({ isFollowing }) => {
+	return (
+		<>
+			{isFollowing && (
```
This might be simpler as:

```tsx
if (isFollowing) {
  return <svg className="following-icon"...>...</svg>
} else {
  return <svg className="follow-icon"...>...</svg>
}
```

But perhaps best not to touch it until merging #1355 to prevent conflicts.

faresite committed:
Merge pull request #1355 from guardian/extract-follow-status

Commit:
refactor

Commit:
update snaps
